main.tf

```diff
@@ -157,3 +158,24 @@ module "database" {
 
   vpc_security_group_ids = [module.vpc.infrastructure_sg_id]
 }
+
+resource "terraform_data" "private_key_copy" {
+  
+for_each = { for host, instance in flatten(module.controller_vm[*].vm_public_ip): host => instance }
```

Comment:
This can be simplified with use of `count`, similar to [this](https://github.com/ansible-content-lab/azure_terraform_deployment/pull/11/files#diff-32130ca832020a58469a9ce227516fe410a494f9967b54ad74df30e3919ed84dR85), so the file provisioner code to copy ssh key will only run based on a condition (app_tag == controller)